Fix sparse bugs; add reduce_add, broadcase, data_parallel for sparse

ssnl opened the pull request:
This PR adds various fixes for sparse tensors. In particular, it contains the following:

Fixes:
1. Fixes `.type()` not converting indices tensor.
2. Fixes ATen not considering sparse cuda tensors as cuda and missing `get_device` bindings, causing `AutoGPU` checks to fail in backward engine.
3. Fixes `AutoGPU` missing args in `pynew`.
4. Fixes sparse tensor coalesce. The bug is caused by using `thrust::unique_by_key`, which modifies the key tensor. In case where key is 1D, original implementation is operating on the same tensor indices data storage, and thus modifies it. This is fixed by adding a flag to `THCSTensor_(newFlattenedIndices)`, controlling whether the returned value is forced to be a clone.

New functionalities:
Added support for sparse tensors in `broadcast_coalesced` and `reduce_add_coalesced`, and thus enabling `data_parallel` (fixing #1456 ). In particular, the `_take_tensors` function is modified:

1.  ~~~In case of `reduce_add_coalesced`, it returns each sparse tensor in a separate chunk because reduce_add on combined tensors only makes sense in dense case. In addition, the yielded tensors are now only maintaining order within sparse/dense class, as specified in code comments. Hence, an additional method `_reorder_tensors_as` is provided to get back the original order.~~~
2. ~~~In case `broadcast_coalesced`, sparse tensors of same type are returned together in chunk. Furthermore, `_flatten_tensors` and `_unflatten_tensors` are updated to support combining sparse tensors, allowing faster broadcasting.~~~

See comment below for details.

Various tests are added for the fixes and new functionalities above.

cc: @ezyang @colesbury

ssnl committed:
sparse broadcast_coalesce and reduce_add_coalesced

ssnl committed:
fix sparse tensor .cpu()

ssnl committed:
test sparse dp, broadcast_coalesced, reduce_add_coalesced

ssnl committed:
fix pynew gpu_guards

ssnl committed:
fix sparse tensor coalesce

ssnl committed:
add tests for the previously failing coalesce case

SsnL commented:
Although it's clearly not the most productive PR in terms of size of code changes, it was great to learn something about ATen, AutoGPU, autograd engine, sparse tensors, cuda-gdb, etc. Feel more comfortable with working on PyTorch now :)

ezyang commented on torch/lib/THCS/generic/THCSTensor.cu:
```diff
@@ -67,7 +64,8 @@ THCSTensor *THCSTensor_(newCoalesce)(THCState *state, THCSTensor *self) {
   cudaStream_t stream = THCState_getCurrentStream(state);
 
-  THCIndexTensor *indices1D = THCSTensor_(newFlattenedIndices)(state, self);
+  // indices will be modified, so we have to clone or use new storage here
```
It doesn't take much more typing to say "modified by Thrust", which will be helpful for anyone else who ever wants to read this code ;)

ezyang commented:
Old PR with some comments for the earlier patchsets at #3178.

gchanan commented:
Can the Aten changes be made in a separate ATen commit?

ezyang commented on torch/lib/THCS/generic/THCSTensor.cu:
```diff
@@ -173,11 +170,17 @@ THCSTensor *THCSTensor_(newCoalesce)(THCState *state, THCSTensor *self) {
 }
 
-THCIndexTensor* THCSTensor_(newFlattenedIndices)(THCState *state, THCSTensor *self) {
+THCIndexTensor* THCSTensor_(newFlattenedIndices)(THCState *state, THCSTensor *self, int forceClone) {
```
A little bit of boolean blindness here, but maybe it's OK.

ezyang commented:
LGTM, but see @gchanan's comment

SsnL commented:
@gchanan Sure! Should they be in a separate commit or a separate PR?

apaszke commented on torch/_utils.py:
```diff
@@ -102,44 +104,107 @@ def _accumulate(iterable, fn=lambda x, y: x + y):
+        if t.is_sparse:
+            outputs.append(next(sparse_tensors))
+        else:
+            outputs.append(next(dense_tensors))
```
A shorter version:
```python
outputs = [next(sparse) if t.is_sparse else next(dense) for t in ordered_tensors]
```

apaszke commented on torch/_utils.py:
```diff
@@ -102,44 +104,107 @@ def _accumulate(iterable, fn=lambda x, y: x + y):
+
+    If sparse_single_chunk=True, yielded tensors are guaranteed to be only of
+    same order within sparse and dense classes, i.e., sparse tensors are of same
+    order with those in input tensors, and same for dense ones.
```
Why would you ever want to have this as `False`? Also the name is misleading compared to the docs. The name seems to indicate that if you set it to `False` then sparse tensors will be chunked, however this really means that we enforce a stricter ordering, right?

SsnL replied:
No. The name is intended to indicate that if it is `False` then the sparse tensors will be chunked. That way `broadcast_coalesced` can benefit from it.

apaszke commented on torch/cuda/comm.py:
```diff
@@ -77,9 +81,11 @@ def reduce_add(inputs, destination=None):
     nccl_root = None
     for i, inp in enumerate(inputs):
         assert inp.is_cuda, "reduce_add expects all inputs to be on GPUs"
+        assert inp.is_sparse == is_sparse, "reduce_add expects inputs to be either all sparse or all dense"
```
I'm not really sure why do we need to enforce this. The function should be able to handle an arbitrary list of tensors

SsnL replied:
I'll remove this.

apaszke commented on torch/cuda/comm.py:
```diff
@@ -120,12 +125,19 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
+        is_sparse = chunks[0][0].is_sparse
+        if is_sparse:
+            tensors = [chunk[0] for chunk in chunks]
+        else:
```
This indexing logic is reaaaaally convoluted. I have no idea what's happening here now. We're checking things 2 levels deep and then we misteriously take the first element from each chunk (I assume it's because they are 1-elem chunks). This piece of code needs more comments

apaszke commented on torch/cuda/comm.py:
```diff
@@ -120,12 +125,19 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
     """
     output = []
-    itrs = [_take_tensors(tensors, buffer_size) for tensors in inputs]
+    itrs = [_take_tensors(tensors, buffer_size, True) for tensors in inputs]
```
It seems to me that weird things can start to happen if different devices produce different configurations of sparse/dense (i.e. buffers will have different sizes and orderings between devices). Also, this function is covered in `is_sparse` conditionals almost entirely. It would be better to scatter lists from different GPUs into sparse/dense buckets (and check that they match in denseness), and do reductions within them independently.

apaszke reviewed (commented):
I think that we shouldn't pretend that we handle sparse and dense tensors in the same way in most of our functions. It seems that a much simpler approach would be to filter them out into two separate lists at the top-level user-accessible functions, and then have separate codepaths for them. `_take_tensors` became horribly complicated, and I'm pretty sure we'll be committing a lot of bugs in the future, because of all these complicated conditionals

gchanan commented:
By an ATen commit, I meant make a PR to ATen directly: https://github.com/zdevito/ATen.

After that's been merged you can ask someone to update ATen in pytorch or create a PR yourself by following these instructions (see the ATen section): https://github.com/pytorch/pytorch/wiki/Updating-TH,-THC,-THNN,-THCUNN-subtrees

killeent commented on torch/lib/ATen/Declarations.cwrap:
```diff
@@ -1296,6 +1296,7 @@
   python_name: get_device
   backends:
     - CUDA
+    - SparseCUDA
```
This has already been merged but I'm not sure this is the best way to do this. Historically we have used the `aten_sparse` flag on a declaration to enable sparse backend. cc @zdevito

SsnL replied:
I tried that but then it also tries to finde `getDevice` on cpu tensors.

SsnL replied:
https://github.com/zdevito/ATen/blob/master/src/ATen/preprocess_declarations.py#L34-L35

killeent replied:
@SsnL yes, it may require changing that code a little bit so that if we've only specified the `CUDA` backend, we only enable the `CUDA` and `SparseCUDA` backends.

SsnL commented:
Okay, this is ready for review. In the updated code, 

1. I rewrote `_take_tensor` so that it groups tensors by types **and return them only in the same order within the same type**. To reorder them, back `_reorder_tensors_as` is also updated. Theoretically performance should improve when relevant methods are called with a mix of tensors of various types.
2. Separated flatten & unflatten behavior on sparse and dense tensors into different methods: `_flatten/unflatten_sparse/dense_tensors` so that they are not doing things like returning different types (tuple vs tensor) basing on inputs. Therefore, the logic to handle sparse/dense difference is naturally moved to the caller of these functions, i.e. `broadcase_coalesced` and `reduce_add_coalesced`. This improves readability since the functions should handle sparse tensors differently.
3. Added some comment about the boolean flag on `THCSTensor_(newFlattenedIndices)`. Clarified that it is Thrust that modifies the index tensor.

ezyang commented on torch/_utils.py:
```diff
@@ -126,20 +146,57 @@ def _unflatten_tensors(flat, tensors):
+def _unflatten_sparse_tensors(flat, tensors):
+    """View flat buffer (containing indices and values) using the sizes of
+    tensors. Assume that tensors are of same sparse type, and that flat is given
+    by _flatten_sparse_tensors.
```
I understand this is a helper function, so the naming/docs are not as important (i.e., "read the source code if you actually want to know what it does"). However, for future reference when writing docstrings, I'd recommend adding: (1) the precise types of all the arguments (e.g., flat is a tuple of (list of dense tensor, list of dense tensor)) and tensor is a list of sparse tensors; (2) the type returned (list of sparse tensors).

SsnL replied:
Thanks Ed. I will add those. :)

ezyang commented:
Great work, thank you!

apaszke commented on torch/cuda/comm.py:
```diff
@@ -119,13 +131,20 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
     output = []
-    itrs = [_take_tensors(tensors, buffer_size) for tensors in inputs]
+    for tensor_at_gpus in zip(*inputs):
+        if tensor_at_gpus[0].is_sparse:
```
This is a bit worrying for me. If, for some reason, GPU 0 will produce a dense result, but GPU 1 will give you sparse grad, you will take the dense code path, and weird things will start happening

apaszke commented on torch/cuda/comm.py:
```diff
@@ -119,13 +131,20 @@ def reduce_add_coalesced(inputs, destination=None, buffer_size=10485760):
-        result = reduce_add(flattened, destination)
-        output.extend(_unflatten_tensors(result, chunks[0]))
-    return tuple(output)
+        tensors = [_flatten_dense_tensors(chunk) for chunk in chunks]
```
It would also be nice to avoid transposing Python data structures 3 times (interpreter overhead), but this seems more complicated.
